Use getfullargspec if available

mdboom opened the pull request:
`inspect.getargspec` raises `DeprecationWarning` on Python 3.5.  It turns out that all of the uses of `getargspec` in `py.test` can be safely replaced with `getfullargspec`, when available.

In `astropy`, we run our tests with `DeprecationWarnings` converted to exceptions to make sure we are on top of any API changes in Python or third-party dependencies.  Since `py.test` now raises `DeprecationWarnings` as a matter of course on Python 3.5, this is a bit of a problem.

The-Compiler commented:
According to [the docs](https://docs.python.org/3.5/library/inspect.html#inspect.getfullargspec), `getfullargspec` is deprecated since 3.5 as well...

So it might be better to switch to signature objects, if available? It's definitely a bad idea to deprecate something and suggest an alternative which isn't available in 2.7... :unamused:

mdboom commented:
Ah, sorry.  I missed that `getfullargspec` was also deprecated (since it, strangely, doesn't raise a `DeprecationWarning` -- and in fact, `getargspec` is implemented as calling `getfullargspec` and then raising a warning...)

I think you're right, though.  We'll probably have to do this in terms of `Signature` where they are available.

The-Compiler commented:
Every warning is only shown once - so if you called `getargspec` first and then `getfullargspec`, you won't see the second warning.

RonnyPfannschmidt commented:
does this change formatting for python3.x?

can we please have a test that demonstrates the behavior to begin with

untitaker commented:
https://twitter.com/DRMacIver/status/643834745051287552

RonnyPfannschmidt commented:
as far as i can tell we have 2 use-cases

* test if a positional argument is taken
* format a argspec as given to the arguments

i believe use-case 2 can be solved better than what we do using the funcsigs backport
and use-case1 should use something like `func.__code__.co_argcount` in a utility function

mdboom commented:
I've just updated this to use signature (when available), though @RonnyPfannschmidt's suggestions seem reasonable as well.  No strong preference, though using `inspect` feels like a more "public API" than `func.__code__.co_argcount`, though I understand both are, strictly speaking, public.

The-Compiler commented on _pytest/python.py:
```diff
@@ -32,6 +32,26 @@
+    def _has_positional_arg(func):
+        sig = inspect.signature(func)
+        params = list(sig.parameters.values())
+        if len(params):
```
You can probably simply do `if params:` here

The-Compiler commented on _pytest/python.py:
```diff
@@ -32,6 +32,26 @@
+        return str(inspect.signature(func))
+else:
+    def _has_positional_arg(func):
+        return inspect.getargspec(func)[0] is not None
```
I think `.getargspec(...)[0]` will never be `None`, it seems to be the empty list instead:

```python
>>> def foo(): pass
... 
>>> inspect.getargspec(foo)
ArgSpec(args=[], varargs=None, keywords=None, defaults=None)
```

I think this should be `bool(inspect.getargspec(func).args)` instead. It's a named tuple since 2.6, so why not use that too. :wink:f

The-Compiler commented:
LGTM otherwise, but I think it'd be nice to have some tests for this - also to make sure the behaviour is always the same

hpk42 commented:
I agree, some testing for the helpers would be good.  Can you do this?  With a bit of luck we can put it into 2.8 which is due to be released any day now.  

FWIW Using ``inspect.signature`` seems fine enough to me and i wouldn't go for using a 3rd party package which I suggest to generally try to avoid unless there is a real big win (e.g. using colorama on windows for terminal writing).

mdboom commented:
I've changed the positional arg test to `__doc__.co_argcount` as suggested (it's definitely more efficient, if nothing else), and added some tests for the argument formatting.

mdboom commented:
Note there is also an instance of this in the vendored `pluggy`.  What's the procedure there?  Make a PR against `pluggy` and then a PR here to revendor it?

hpk42 commented:
great, thanks @mdboom! yes, if you can do a little PR against pluggy then we can merge and release there and vendor it back.

hpk42 commented on testing/python/metafunc.py:
```diff
@@ -21,6 +21,7 @@ def function(): pass
         metafunc = self.Metafunc(function)
         assert not metafunc.fixturenames
         repr(metafunc._calls)
+        assert funcargs._format_args(function) == '()'
```
please put the tests of the helpers (also below) rather into dedicated test functions.  Tucking it on existing functions like this mixes things too much.

mdboom replied:
Done.

RonnyPfannschmidt commented:
@hpk42 we should consider moving this to a common library when handling py.code as well
should someone create a task issue about that?
